### PR TITLE
Audio service

### PR DESCRIPTION
**Pull request by dave-ryley**
Added the audio_service and config_service as our first services in the game! I think it would be really good to go through this with both you guys and do a bit of a show and tell. I'll be around tomorrow evening if either of you want to do a review.

### CODE REVIEW

File: BOOM/Code/Mixins/character_audio_mixin.lua [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        audio.soundData[k] = v
+    end
+
+    o:addEventListener("onFootstep", audio)
+    o:addEventListener("onDeath", audio)
+    o:addEventListener("onAttack", audio)
+    o.audio = audio
+end
+
+return CharacterAudio
```
Comment (david159ryan) [inline BOOM/Code/Mixins/character_audio_mixin.lua:62]: no newline, fix newlines on all files

File: BOOM/Code/Utilities/global_utils.lua [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    str = str .. "}\n"
+    return str
+end
+
+------------------------------------------------------------------------------------
+-- Table Functions
+------------------------------------------------------------------------------------
+
+function clear(t)
+    for k, v in pairs(t) do
```
Comment (david159ryan) [inline BOOM/Code/Utilities/global_utils.lua:39]: unused variable `v`, replace with `_`

File: BOOM/Code/Utilities/global_utils.lua [hunk 3]
```diff
@@ -0,0 +1,74 @@
+------------------------------------------------------------------------------------
+-- Local Functions
+------------------------------------------------------------------------------------
+
+local function addWhiteSpace(str, depth)
+    for i = 1, depth do
```
Comment (david159ryan) [inline BOOM/Code/Utilities/global_utils.lua:6]: unused variable i, replace with _

File: BOOM/Code/Services/config_service.lua [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+
+------------------------------------------------------------------------------------
+-- Local Functions - Loading
+------------------------------------------------------------------------------------
+
+local namespaceCheck = {}
+local function load(requirePath)
+    -- print("Loading Config:", requirePath)
+    local configFile = require(requirePath)
+
```
Comment (david159ryan) [inline BOOM/Code/Services/config_service.lua]: whitespace

File: BOOM/Code/Services/config_service.lua [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+    return config
+end
+
+function ConfigService:buildAssetTable(namespaceKey, assetKey, loadFunction, result)
+    local namespace = Configs[namespaceKey]
+    if not namespace then
+        error("Cannot find namespace \"" .. namespaceKey .. "\" in Configs. Cannot build Asset Table.")
+        return result
+    end
+    return find
```
Comment (david159ryan) [inline BOOM/Code/Services/config_service.lua]: undefined variable, is this intentional?

File: BOOM/Code/Services/config_service.lua [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+        error("Invalid Id:" .. configId .. ". Could not find namespace: " .. splitId[1])
+        return nil
+    end
+
+    local config = getConfig(splitId, 2, namespace)
+    clear(splitId)
+    return config
+end
+
+function ConfigService:buildAssetTable(namespaceKey, assetKey, loadFunction, result)
```
Comment (david159ryan) [inline BOOM/Code/Services/config_service.lua]: `assetKey` and `loadFunction` unused

File: BOOM/Code/Services/audio_service.lua [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+                channel = c,
+                loops = 0,
+                fadein = 0,
+            }
+        )
+    end
+end
+
+function AudioService:playMusic(configId)
+    local asset = MusicConfig[configId]
```
Comment (david159ryan) [inline BOOM/Code/Services/audio_service.lua]: `MusicConfig` is undefined

File: BOOM/Config/Audio/music_config.lua [hunk 8]
```diff
@@ -1,7 +1,28 @@
...
-    ["LEVEL_2"] = "Resources/Sounds/Music/DeathCell.ogg",
-    ["LEVEL_3"] = "Resources/Sounds/Music/HeadShredder.ogg",
-}
+return
+{
+    _namespace = "MUSIC",
 
-return Music
\ No newline at end of file
+    BASE =
```
Comment (david159ryan) [inline BOOM/Config/Audio/music_config.lua:5]: trailing whitespace in this file

**Review by david159ryan — CHANGES_REQUESTED** [review]
Setup your editor to trim trailing while spaces and add a newline at the end of a file. Also it looks like some of this is still WIP (undefined variables and unused fucntion arguments), so maybe it's best to not merge features to master until they are complete. Apart from that it looks great!

File: BOOM/Code/Services/config_service.lua (hunk 5 above)
Comment (dave-ryley) [reply BOOM/Code/Services/config_service.lua]: Ya that was a bug, it was supposed to be calling the recursive findAndLoadAsset() function..

File: BOOM/Code/Services/config_service.lua (hunk 6 above)
Comment (dave-ryley) [reply BOOM/Code/Services/config_service.lua]: Same as above. Was supposed to pass to the findAndLoadAsset() function

File: BOOM/Code/Services/audio_service.lua (hunk 7 above)
Comment (dave-ryley) [reply BOOM/Code/Services/audio_service.lua]: Fixed this. Was using the old method of Config loading before I created the ConfigService

**Comment by dave-ryley** [comment]
Fixed the whitespace. Fixed the other stuff.
There is some WIP code in there still, but only on the CharacterAudioMixin, which isn't yet being used. I need to set up an EventMixin for the game before I can use it. (which is why the o:addEventListener() functions wouldn't run right now). But CharacterAudioMixin isn't required anywhere yet, so it should be fine to push.

**Comment by david159ryan** [comment]
I would say best to finish the feature before merging, there is no benefit to having half implemented code in master. Especially if there's a chance of merge conflicts. Best to get them out of the way in one go.

**Comment by dave-ryley** [comment]
CharacterAudioMixin is now relying on the Event Mixin branch, if both of these are merged into master there should be no issues using CharacterAudioMixin on an entity that has an EventManagerMixin

**Pull request merged**